[WIP] OpenTofu support

meringu opened the pull request:
## what

This is an early attempt at getting OpenTofu to work with Atlantis. I thought I'd get open this PR sooner rather than later, as conflicts with #4494 and would be interested to see what people think or get any feedback.

I'm just starting to test this internally and have made some quick decisions in order to get it working:
- Introduces a `--tf-distribution` setting that can be set to `terraform` or `opentofu`.
- OpenTofu is downloaded through GitHub releases from https://github.com/opentofu/opentofu
  - This method is not compatible with releases.hashicorp.com, so currently ignores `--tf-download-url`
- OpenTofu ships the archives with additional files, just like Terraform 1.8.2. Therefore this change should also resolve the need for #4483.
- Added an interface to `server/core/terraform/terraform_client.go` for the Terraform distribution. This matches up with the `--tf-distribution` setting. I'm not sure if this is the right place or the right abstraction, but seemed to be a relatively low effort place to implement it.
- I would like to also support configuring Terraform or OpenTofu per project in the server side config or repo level config. This would support an easier path to migrate between the two.
- There is a small change to set a default TARGETPLATFORM in the Dockerfile. This is to support kaniko, which I'm using internally to build and test this. Happy to revert that.
- I'm not checking the signature of the opentofu archive yet.

I've done limited testing so far, but would like to get some feedback before I continue down the wrong rabbit hole.

## why

#3741

## tests

I've got them working for the new download method, but should also add some for the new OpenTofu code paths.

## references

#3741

james0209 commented:
I agree with the premise of this, and I like the solution, but I don't agree that it stops the need for hc-install - I think that is worthy of a discussion.

I do not think your changes  inherently conflict with the hc-install PR - it is very easy to do this for example

```
func (c *DefaultClient) DetectVersion(log logging.SimpleLogging, projectDirectory string) *version.Version {
	return c.distribution.DetectVersion(log, c, projectDirectory)
}

type Distribution interface {
	BinName() string
	SourceURL(v *version.Version, downloadURL string) string
	DetectVersion(log logging.SimpleLogging, c *DefaultClient, projectDirectory string) *version.Version
}

func (*DistributionTerraform) DetectVersion(log logging.SimpleLogging, c DefaultClient, projectDirectory string) *version.Version {
// use hc-install
}

func (*DistributionOpenTofu) ListAvailableVersions(log logging.SimpleLogging, downloadBaseURL string, downloadAllowed bool) ([]string, error) {
// use current implementation - only called by the DetectVersion below
}

func (dt *DistributionOpenTofu) DetectVersion(log logging.SimpleLogging, c DefaultClient, projectDirectory string) *version.Version {
// use current implementation
}
```

Especially as the `ListAvailableVersions` func is only called in the current implementation of DetectVersion - there is no other usage.

I don't mind if the hc-install approach is not chosen, but I think if you are going half of the way to have different interfaces and methods for OpenTofu vs Terraform implementation, then a discussion on whether to use the same methods vs different methods for each Distribution should be had.

meringu commented:
Thanks @james0209, that makes sense.

I'd be happy to rebase this over the hc-install change. As you point out, that should be fairly straightforward.

nitrocode commented:
Due to wip status, I switched this to a draft as a non draft means that it's ready for review. Please set it as ready to review when ready. 

Thank you for the contribution

chenrui333 commented:
> I'd be happy to rebase this over the hc-install change. As you point out, that should be fairly straightforward.

we have merged hc-install update, might worth do the rebase and pick up this PR again. Thanks for all the efforts!

jamengual commented:
The OpenTofu team is going to release this downloader to make it easier for us to auto download Opentofu.

https://github.com/janosdebugs/downloader
You can pull it in with a replace statement in go.mod for now until is released today or Monday.

@meringu @nitrocode if any of you want to work on this let us know

meringu commented:
Awesome, I rebased this from main yesterday and got to the part where we need to download tofu.

I'm keen to keep working on this, but happy to have it taken over if I'm being too slow.

meringu commented:
Tofu downloader has been released: https://github.com/opentofu/tofudl

meringu commented:
Thanks for your patience. I've done the following:
- Merged from main to get the hc-install changes.
- Updated to use tofu downloader.
- Refactored out a distribution and downloader.
- Made a separate downloader for conftest to decouple it from the TF downloaders.

I've done some testing on one of my Atlantis servers. I've been able to do successful plan and applies. Have also tested resolving different version from constraints from the `required_version` setting.

meringu commented:
Thanks @X-Guardian, I have cherry picked out a new commit, and only run `go generate` on the packages I modified.

meringu commented:
Have updated the PR description to reflect the changes. The PR is labeled as needs-discussion and waiting-on-response, so please let me know how I can help.

jamengual commented:
@meringu can you resolve the conflicts and if is ready for review, can you please change it to non-draft? Thanks.

GenPage commented on server/core/terraform/terraform_client.go:
```diff
@@ -128,17 +119,19 @@ func NewClientWithDefaultVersion(
 	if err == nil {
 		localVersion, err = getVersion(localPath)
 		if err != nil {
+			panic(err)
```
Do we want to panic here? I do not want to introduce new panics to the codebase without good reason.

GenPage commented on runatlantis.io/docs/server-configuration.md:
```diff
@@ -1266,6 +1274,8 @@ Setting this to `false` can be useful in an air-gapped environment where a downl
 
   This has no impact if `--tf-download` is set to `false`.
 
+  This setting is currently ignored if `--tf-distribution` is set to `opentofu`
```
What is the rationale behind ignoring this for opentofu? Does this mean we always download the binary?

meringu replied:
I'm not sure exaclty how to support this flag with opentofu. I'm open to ideas.

Tofudl supports two fields to download from a mirror: https://github.com/opentofu/tofudl/blob/76c6857d6eac665824b29da4e893e06e0b1896b2/branding/branding.go#L10-L14. I think we'd need to add an additional flag at the very least? I'm also not sure how specific this setting is to tofudl, that might be an issue if we change download method.

jamengual replied:
@meringu we can maybe ask on the opentofu slack or an issue to tofudl if neccesary

meringu replied:
Great idea, I have asked over here: https://opentofucommunity.slack.com/archives/C05RKE7PS72/p1726523077638359

meringu replied:
The thinking is that we could look at supporting this later once someone needs it and has defined use-case. For now, I've updated the wording to indicate that it is not yet supported, and can unblock us getting this in for others to start to try out OpenTofu without a mirror.

GenPage commented on server/core/terraform/downloader.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	// Write out the tofu binary to the disk:
+	file := filepath.Join(dir, "terraform"+v.String())
```
writing the tofu binary as "terraform." Is this intentional to avoid refactoring more of the codebase even though we track the name via distribution.go?

GenPage reviewed (dismissed):
Thank you for taking the time to implement this feature. A few comments before approving.

meringu commented:
Looks like tofudl is failing to compile on 32bit as this constant is too large: https://github.com/opentofu/tofudl/blob/76c6857d6eac665824b29da4e893e06e0b1896b2/branding/branding.go#L36

I'll looks at raising a PR.

meringu commented on server/core/terraform/distribution.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+func (*DistributionOpenTofu) ResolveConstraint(constraintStr string) (*version.Version, error) {
+	dl, err := tofudl.New()
```
Based on the discussion here https://github.com/runatlantis/atlantis/pull/4499#discussion_r1720373798, and in the Slack thread you mentioned that you weren't aware of any mirrors yet, my preference is to park support for that until we need it.

I also double checked the issue for adding OpenTofu to Atlantis, and no one has requested mirror support yet.

I'll resolve this, and we can use the discussion above.

lukemassa reviewed (approved):
I tested locally and it seems to work fine, and the code looks good to me!

Thanks for the contribution!

jamengual commented:
Thanks @meringu and everyone that helped on this.
